src/builtins/impls/urlquery.rs

```diff
@@ -16,28 +16,70 @@
 
 use std::collections::HashMap;
 
-use anyhow::{bail, Result};
+use serde_json::Value;
+use url::form_urlencoded::{byte_serialize, parse};
+
+const QUERY_CONCAT_CHAR: &str = "&";
 
 /// Decodes a URL-encoded input string.
-#[tracing::instrument(name = "urlquery.decode", err)]
-pub fn decode(x: String) -> Result<String> {
-    bail!("not implemented");
+#[tracing::instrument(name = "urlquery.decode")]
+pub fn decode(x: String) -> String {
+    parse(x.as_bytes())
+        .map(|(key, val)| [key, val].concat())
+        .collect()
 }
 
 /// Decodes the given URL query string into an object.
-#[tracing::instrument(name = "urlquery.decode_object", err)]
-pub fn decode_object(x: String) -> Result<HashMap<String, Vec<String>>> {
-    bail!("not implemented");
+#[tracing::instrument(name = "urlquery.decode_object")]
+pub fn decode_object(x: String) -> HashMap<String, Vec<String>> {
+    let mut decoded_object: HashMap<String, Vec<String>> = HashMap::new();
+    for pair in x.split(QUERY_CONCAT_CHAR) {
+        let mut iter_parameter = pair.split('=').take(2);
+        let (parameter_key, parameter_value) = match (iter_parameter.next(), iter_parameter.next())
+        {
+            (Some(k), Some(v)) => (k, v),
+            _ => continue,
+        };
+        match decoded_object.get_mut(parameter_key) {
+            Some(v) => v.push(parameter_value.to_string()),
+            _ => {
+                decoded_object.insert(parameter_key.to_string(), vec![parameter_value.to_string()]);
+            }
+        };
+    }
+    decoded_object
 }
 
 /// Encodes the input string into a URL-encoded string.
-#[tracing::instrument(name = "urlquery.encode", err)]
-pub fn encode(x: String) -> Result<String> {
-    bail!("not implemented");
+#[tracing::instrument(name = "urlquery.encode")]
+pub fn encode(x: String) -> String {
+    byte_serialize(x.as_bytes()).collect()
 }
 
 /// Encodes the given object into a URL encoded query string.
-#[tracing::instrument(name = "urlquery.encode_object", err)]
-pub fn encode_object(x: HashMap<String, serde_json::Value>) -> Result<String> {
-    bail!("not implemented");
+#[tracing::instrument(name = "urlquery.encode_object")]
+pub fn encode_object(x: HashMap<String, serde_json::Value>) -> String {
```

Comment:
Instead of implementing that manually, you should use `form_urlencoded::Serializer`.

It works like that:

```rust
let encoded: String = form_urlencoded::Serializer::new(String::new())
    .append_pair("foo", "bar & baz")
    .append_pair("saison", "Été+hiver")
    .finish();
```

so I'd do that in a loop.

The other thing is, instead of accepting a `HashMap<String, serde_json::Value>`, I'd accept explicitly a map of either strings or list of strings as parameter. Something like this:


```rust
#[derive(Serialize, Deserialize)]
#[serde(untagged)]
enum OneOrMany<T> {
    One(T),
    Many(Vec<T>),
}

pub fn encode_object(x: HashMap<String, OneOrMany<String>>) -> String {
  // ...
}
```

Reply:
changed this function to use `form_urlencoded::Serializer`.
When i tried to changed the `pub fn encode_object(x: HashMap<String, OneOrMany<String>>)` i got.

regarding the `OneOrMany` enum, not sure how to doing it